src/Mvc/Mvc.RazorPages/src/Infrastructure/PageActionEndpointDataSource.cs

```diff
@@ -29,6 +33,13 @@ internal class PageActionEndpointDataSource : ActionEndpointDataSourceBase
             _orderSequence = orderedEndpoints;
             DefaultBuilder = new PageActionEndpointConventionBuilder(Lock, Conventions);
 
+            // If we haven't replaced the default view compiler then we can load the page directly
+            // synchronously and in memory
+            if (pageLoader is DefaultPageLoader pl && pl.Compiler is DefaultViewCompiler)
```

Comment:
We might be able to take this further and say the loader does not get used by the data source unless runtime compilation is in play. This would let us 

a) Simplify this data source to read directly from the view feature and populate ADs
b) Move the current implementation of the data source + matcher policy to runtime compilation (or have it continue being here and you have to use an option or an API to enable it).

Reply:
Right, I didn't go that far because we still need to run this logic:

https://github.com/dotnet/aspnetcore/blob/19c492ffca19dd25ace7290cf6cf0a0cebff5be2/src/Mvc/Mvc.RazorPages/src/Infrastructure/DefaultPageLoader.cs#L70-L83

Reply:
We can definitely move the matcher policy to the runtime compilation assembly though. We can also treat the IViewCompiler abstraction like a runtime only thing. Is that what you're thinking?